compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt

```diff
@@ -29,11 +30,14 @@ object ArrayIterator : IntrinsicMethod() {
         signature: JvmMethodSignature,
         context: JvmBackendContext
     ): IrIntrinsicFunction {
-        val owner = context.typeMapper.mapClass(expression.symbol.owner.parentAsClass)
+        val isPrimitive = ((expression.type as? IrSimpleType)?.arguments?.size ?: 0) == 0
```

Comment:
This change can break other back-ends, we should at least check it by remote run (I will do). In general, it's better to avoid changes in back-ends when fixing FIR2IR. I think @udalov can tell us more precise whether this change is acceptable or not.

Reply:
I agree. It's a bit suspicious in general to adapt the JVM IR backend to two different frontend models. I'd strongly prefer to keep the frontend model (or at least the part that the backend works with) the same between psi2ir and fir2ir.

In this case, the change looks safe but I'm very curious why is it needed, because it may reveal deeper problems in fir2ir.

Reply:
Updated the commit to change fir2ir, so that ArrayIterator can stay as it is.

It seems the backend has some baked-in knowledge of how psi2ir translated to IR, which is often not followed by fir2ir. I'll make fir2ir follow what's expected by the backend, to avoid changing the backend.